### PR TITLE
Charlesmchen/batch message processing

### PR DESCRIPTION
**Pull request by charlesmchen**
Batch decrypt and process of incoming messages.

* There's a significant perf win to processing messages in batches, not least because `YapDatabaseModifiedNotification` is only fired once per batch.
* There's a similar win even when only processing single messages by virtue of using a single transaction, because message processing involves a lot of database reads and writes... and we'll also only fire the "database modified" notification once per message.

PTAL @michaelkirk

### CODE REVIEW

File: SignalServiceKit/src/Messages/TSMessagesManager.m [hunk 1]
```diff
@@ -593,54 +628,65 @@ - (void)handleIncomingEnvelope:(OWSSignalServiceProtosEnvelope *)incomingEnvelop
...
     }
 
-    if (callMessage.hasOffer) {
-        [self.callMessageHandler receivedOffer:callMessage.offer fromCallerId:incomingEnvelope.source];
-    } else if (callMessage.hasAnswer) {
-        [self.callMessageHandler receivedAnswer:callMessage.answer fromCallerId:incomingEnvelope.source];
-    } else if (callMessage.iceUpdate.count > 0) {
-        for (OWSSignalServiceProtosCallMessageIceUpdate *iceUpdate in callMessage.iceUpdate) {
-            [self.callMessageHandler receivedIceUpdate:iceUpdate fromCallerId:incomingEnvelope.source];
+    // TODO: Should we do this synchronously?
```
Comment (charlesmchen) [inline SignalServiceKit/src/Messages/TSMessagesManager.m]: Significant question.  We have two options:

* If we want to dispatch these call messages to the `callMessageHandler` synchronously, we'll need to plumb the transaction through to the call service which is non-trivial. This will let us ensure that the "unprocessed message queue" managed by `OWSBatchMessageProcessor` isn't evacuated until the call message is received.
* We can do what I've done here which is a lot simpler... call messages will only be dropped if the app happens to exit before the async dispatch occurs, which is fine - if the app exits, the call will end.

**Comment by charlesmchen** [comment]
It's worth noting that another huge change here is that most message handling (both decrypt and "processing" steps) is now done off the main thread.

File: Signal/src/ViewControllers/OWSNavigationController.m [hunk 2]
```diff
@@ -51,8 +51,9 @@ - (BOOL)navigationBar:(UINavigationBar *)navigationBar shouldPopItem:(UINavigati
     // If we're not going to cancel the pop/back, we need to call the super
     // implementation since it has important side effects.
     if (result) {
+        // NOTE: result might end up NO if the super implementation cancels the
+        //       the pop/back.
         result = [super navigationBar:navigationBar shouldPopItem:item];
-        OWSAssert(result);
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/OWSNavigationController.m]: This is a small unrelated fix.

File: SignalServiceKit/src/Messages/OWSBatchMessageProcessor.m [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+#import <YapDatabase/YapDatabaseTransaction.h>
+#import <YapDatabase/YapDatabaseView.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+#pragma mark - Persisted data model
+
+@class OWSSignalServiceProtosEnvelope;
+
+@interface OWSBatchMessageProcessingJob : TSYapDatabaseObject
```
Comment (charlesmchen) [inline SignalServiceKit/src/Messages/OWSBatchMessageProcessor.m]: I've closed followed the pattern you set up in the message receiver.

File: SignalServiceKit/src/Messages/OWSMessageDecrypter.m [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+@property (nonatomic, readonly) TSStorageManager *storageManager;
+@property (nonatomic, readonly) YapDatabaseConnection *dbConnection;
+@property (nonatomic, readonly) OWSBlockingManager *blockingManager;
+@property (nonatomic, readonly) OWSIdentityManager *identityManager;
+
+@end
+
+#pragma mark -
+
+@implementation OWSMessageDecrypter
```
Comment (charlesmchen) [inline SignalServiceKit/src/Messages/OWSMessageDecrypter.m]: I've pulled out the decryption logic from the message manager into a separate class.

File: SignalServiceKit/src/Messages/OWSMessageHandler.h [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+extern const NSUInteger kIncomingMessageBatchSize;
+
+@class OWSSignalServiceProtosContent;
+@class OWSSignalServiceProtosDataMessage;
+@class OWSSignalServiceProtosEnvelope;
+
+NSString *envelopeAddress(OWSSignalServiceProtosEnvelope *envelope);
+
+@interface OWSMessageHandler : NSObject
```
Comment (charlesmchen) [inline SignalServiceKit/src/Messages/OWSMessageHandler.h]: A common base class for the "message decrypter" and the "message manager".  Contains utility methods used for logging.

File: SignalServiceKit/src/Messages/OWSMessageManager.h [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+#import "OWSMessageHandler.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@class OWSSignalServiceProtosEnvelope;
+@class TSThread;
+@class YapDatabaseReadWriteTransaction;
+
+@interface OWSMessageManager : OWSMessageHandler
```
Comment (charlesmchen) [inline SignalServiceKit/src/Messages/OWSMessageManager.h]: Renamed `TSMessagesManager` to `OWSMessageManager` and cleaned up its header.

**Comment by charlesmchen** [comment]
This PR was getting hard to read so I've decomplosed it into two PRs.  This first half contains the work to batch process incoming messages.  

The second PR https://github.com/WhisperSystems/Signal-iOS/pull/2532 refactors the message processing code, cleans up interfaces and imports, renames classes, etc.

File: SignalServiceKit/src/Devices/OWSReadReceiptsProcessor.m [hunk 7]
```diff
@@ -80,10 +80,20 @@ - (instancetype)initWithIncomingMessage:(TSIncomingMessage *)message storageMana
 
 - (void)process
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalServiceKit/src/Devices/OWSReadReceiptsProcessor.m]: I'm considering reworking `OWSReadReceiptsProcessor` significantly, folding it into the new `OWSReadReceiptManager`, so I'd rather not refactor it here.

File: SignalServiceKit/src/Contacts/Threads/TSGroupThread.m [hunk 8]
```diff
@@ -162,12 +176,23 @@ - (NSString *)name
 
 - (void)updateAvatarWithAttachmentStream:(TSAttachmentStream *)attachmentStream
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalServiceKit/src/Contacts/Threads/TSGroupThread.m]: I'm ambivalent - I definitely am sympathetic to your motive, but I'm not sure its practical to remove all new transactions from our models since this is so pervasive in our codebase.  And it would involve a lot more verbosity.  

And if it's alright with you, I'd rather not make those changes in this PR which is already huge. Let's discuss.

File: SignalServiceKit/src/Messages/Attachments/OWSAttachmentsProcessor.h [hunk 9]
```diff
@@ -31,17 +33,24 @@ extern NSString *const kAttachmentDownloadAttachmentIDKey;
...
 
 /*
  * Retry fetching failed attachment download
  */
 - (instancetype)initWithAttachmentPointer:(TSAttachmentPointer *)attachmentPointer
-                           networkManager:(TSNetworkManager *)networkManager NS_DESIGNATED_INITIALIZER;
+                           networkManager:(TSNetworkManager *)networkManager
+                           storageManager:(TSStorageManager *)storageManager NS_DESIGNATED_INITIALIZER;
 
 - (void)fetchAttachmentsForMessage:(nullable TSMessage *)message
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalServiceKit/src/Messages/Attachments/OWSAttachmentsProcessor.h]: See above - I'd like to merge this class with another.

File: SignalServiceKit/src/Messages/Interactions/TSIncomingMessage.m [hunk 10]
```diff
@@ -75,37 +75,39 @@ - (instancetype)initWithTimestamp:(uint64_t)timestamp
...
+        enumerateMessagesWithTimestamp:timestamp
+                             withBlock:^(NSString *collection, NSString *key, BOOL *stop) {
+                                 TSInteraction *interaction =
+                                     [TSInteraction fetchObjectWithUniqueID:key transaction:transaction];
+                                 if ([interaction isKindOfClass:[TSIncomingMessage class]]) {
+                                     TSIncomingMessage *message = (TSIncomingMessage *)interaction;
+
+                                     // Only groupthread sets authorId, thus this crappy code.
+                                     // TODO ALL incoming messages should have an authorId.
+                                     NSString *messageAuthorId;
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalServiceKit/src/Messages/Interactions/TSIncomingMessage.m]: I've updated all instances of this comment.

File: SignalServiceKit/src/Messages/OWSBatchMessageProcessor.m [hunk 11]
```diff
@@ -0,0 +1,435 @@
...
+    return self;
+}
+
+- (NSArray<OWSBatchMessageProcessingJob *> *)nextJobsForBatchSize:(NSUInteger)maxBatchSize
+{
+    NSMutableArray<OWSBatchMessageProcessingJob *> *jobs = [NSMutableArray new];
+    [self.dbConnection readWithBlock:^(YapDatabaseReadTransaction *_Nonnull transaction) {
+        YapDatabaseViewTransaction *viewTransaction = [transaction ext:OWSBatchMessageProcessingJobFinderExtensionName];
+        OWSAssert(viewTransaction != nil);
+        NSMutableArray<NSString *> *jobIds = [NSMutableArray new];
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalServiceKit/src/Messages/OWSBatchMessageProcessor.m]: Done.

File: SignalServiceKit/src/Messages/OWSBatchMessageProcessor.m [hunk 12]
```diff
@@ -0,0 +1,435 @@
...
+                [OWSBatchMessageProcessingJob fetchObjectWithUniqueID:jobId transaction:transaction];
+            if (job) {
+                [jobs addObject:job];
+            } else {
+                OWSFail(@"Could not load job: %@", jobId);
+            }
+        }
+    }];
+
+    return jobs;
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalServiceKit/src/Messages/OWSBatchMessageProcessor.m]: Done.

File: SignalServiceKit/src/Messages/OWSBatchMessageProcessor.m [hunk 13]
```diff
@@ -0,0 +1,435 @@
...
+{
+    [self.dbConnection readWriteWithBlock:^(YapDatabaseReadWriteTransaction *_Nonnull transaction) {
+        [[[OWSBatchMessageProcessingJob alloc] initWithEnvelopeData:envelopeData plaintextData:plaintextData]
+            saveWithTransaction:transaction];
+    }];
+}
+
+- (void)removeJobWithId:(NSString *)uniqueId
+{
+    [self.dbConnection readWriteWithBlock:^(YapDatabaseReadWriteTransaction *_Nonnull transaction) {
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalServiceKit/src/Messages/OWSBatchMessageProcessor.m]: Done.

File: SignalServiceKit/src/Messages/OWSMessageReceiver.m [hunk 14]
```diff
@@ -185,18 +216,23 @@ + (void)syncRegisterDatabaseExtension:(YapDatabase *)database
 @interface OWSMessageProcessingQueue : NSObject
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalServiceKit/src/Messages/OWSMessageReceiver.m]: Done, good point.  Note: I'm going to continue to use the same YDB collection names, etc. to prevent data loss.

File: SignalServiceKit/src/Messages/OWSMessageReceiver.m [hunk 15]
```diff
@@ -222,43 +274,95 @@ - (void)enqueueEnvelopeForProcessing:(OWSSignalServiceProtosEnvelope *)envelope
...
+    plaintextDataMap:(NSMutableDictionary<NSString *, NSData *> *)plaintextDataMap
+          completion:(void (^)())completion
+{
+    OWSAssert(jobs.count > 0);
+    OWSAssert(unprocessedJobs.count <= jobs.count);
+
+    if (unprocessedJobs.count < 1) {
+        for (OWSMessageProcessingJob *job in jobs) {
+            NSData *_Nullable plaintextData = plaintextDataMap[job.uniqueId];
+            [self.batchMessageProcessor enqueueEnvelopeData:job.envelopeData plaintextData:plaintextData];
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalServiceKit/src/Messages/OWSMessageReceiver.m]: Current batch size is 10, current max envelope size is 250k, decrypted size should be smaller than encrypted size, so max batch size in memory < 2.5 MB.  I thought that was reasonable?  There are some perf benefits to this arrangement.

[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalServiceKit/src/Messages/OWSMessageReceiver.m]: Note that most batches will be just 1-2 messages and batches should be short-lived in memory.  If we're doing large batches its because "we're slammed" for one reason or another and perf will matter.

File: SignalServiceKit/src/Messages/OWSMessageSender.h [hunk 16]
```diff
@@ -67,6 +68,10 @@ NS_SWIFT_NAME(MessageSender)
 - (void)sendMessage:(TSOutgoingMessage *)message
             success:(void (^)())successHandler
             failure:(void (^)(NSError *error))failureHandler;
+- (void)sendMessage:(TSOutgoingMessage *)message
+        transaction:(YapDatabaseReadWriteTransaction *_Nullable)transaction
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalServiceKit/src/Messages/OWSMessageSender.h]: Done.

File: SignalServiceKit/src/Messages/TSMessagesManager.m [hunk 17]
```diff
@@ -260,84 +269,90 @@ - (NSString *)descriptionForSyncMessage:(OWSSignalServiceProtosSyncMessage *)syn
...
+#pragma mark - Blocking
 
-- (void)processEnvelope:(OWSSignalServiceProtosEnvelope *)envelope
-             completion:(nullable MessageManagerCompletionBlock)completionHandler
+- (BOOL)isEnvelopeBlocked:(OWSSignalServiceProtosEnvelope *)envelope
 {
-    OWSAssert([NSThread isMainThread]);
+    OWSAssert(envelope);
+
+    return [_blockingManager.blockedPhoneNumbers containsObject:envelope.source];
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalServiceKit/src/Messages/TSMessagesManager.m]: Okay, but I'm not applying inside loops since its not free.

File: SignalServiceKit/src/Messages/TSMessagesManager.m [hunk 18]
```diff
@@ -347,134 +362,172 @@ - (void)processEnvelope:(OWSSignalServiceProtosEnvelope *)envelope
...
-                        [NSString stringWithFormat:@"Exception while decrypting: %@", exception.description];
-                    NSError *error = OWSErrorWithCodeDescription(OWSErrorCodeFailedToDecryptMessage, errorDescription);
-                    completion(error);
-                });
-                return;
-            }
+    [self decryptEnvelope:envelope
+        messageTypeName:@"PreKey Bundle"
+        missingPayloadBlock:^{
+            OWSProdFail([OWSAnalyticsEvents messageManagerErrorPrekeyBundleEnvelopeHasNoContent]);
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalServiceKit/src/Messages/TSMessagesManager.m]: Done.

File: SignalServiceKit/src/Messages/TSMessagesManager.m [hunk 19]
```diff
@@ -347,134 +362,172 @@ - (void)processEnvelope:(OWSSignalServiceProtosEnvelope *)envelope
...
+        failureBlock:failureBlock];
+}
+
+- (void)decryptEnvelope:(OWSSignalServiceProtosEnvelope *)envelope
+        messageTypeName:(NSString *)messageTypeName
+    missingPayloadBlock:(void (^_Nonnull)())missingPayloadBlock
+     cipherMessageBlock:(id<CipherMessage> (^_Nonnull)(NSData *))cipherMessageBlock
+           successBlock:(DecryptSuccessBlock)successBlock
+           failureBlock:(void (^)(NSError *_Nullable error))failureBlock
+{
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalServiceKit/src/Messages/TSMessagesManager.m]: > we could take in an instantiated (id<CipherMessage>) cipherMessage

That would require adding "setData" methods to these cipher message classes. I'd rather not touch `AxolotlKit`.

File: SignalServiceKit/src/Messages/TSMessagesManager.m [hunk 20]
```diff
@@ -347,134 +362,172 @@ - (void)processEnvelope:(OWSSignalServiceProtosEnvelope *)envelope
...
+    OWSAssert(envelope);
+    OWSAssert(messageTypeName.length > 0);
+    OWSAssert(missingPayloadBlock);
+    OWSAssert(cipherMessageBlock);
+    OWSAssert(successBlock);
+    OWSAssert(failureBlock);
+
+    TSStorageManager *storageManager = [TSStorageManager sharedManager];
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalServiceKit/src/Messages/TSMessagesManager.m]: Done.

File: SignalServiceKit/src/Messages/TSMessagesManager.m [hunk 21]
```diff
@@ -347,134 +362,172 @@ - (void)processEnvelope:(OWSSignalServiceProtosEnvelope *)envelope
...
+    NSString *recipientId = envelope.source;
+    int deviceId = envelope.sourceDevice;
+
+    // DEPRECATED - Remove after all clients have been upgraded.
+    NSData *encryptedData = envelope.hasContent ? envelope.content : envelope.legacyMessage;
+    if (!encryptedData) {
+        missingPayloadBlock();
+        failureBlock(nil);
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalServiceKit/src/Messages/TSMessagesManager.m]: Not a bad idea.  I've made a card. https://trello.com/c/gEvMfngy/1561-always-use-explicit-errors-in-message-decrypt-and-message-processing

File: SignalServiceKit/src/Messages/TSMessagesManager.m [hunk 22]
```diff
@@ -347,134 +362,172 @@ - (void)processEnvelope:(OWSSignalServiceProtosEnvelope *)envelope
...
+                                                                    preKeyStore:storageManager
+                                                              signedPreKeyStore:storageManager
+                                                               identityKeyStore:self.identityManager
+                                                                    recipientId:recipientId
+                                                                       deviceId:deviceId];
+
+            NSData *plaintextData = [[cipher decrypt:cipherMessage] removePadding];
+            successBlock(plaintextData);
+        } @catch (NSException *exception) {
             dispatch_async(dispatch_get_main_queue(), ^{
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalServiceKit/src/Messages/TSMessagesManager.m]: I'd love to move more off the main thread here, but I'll take this on later to avoid rebase hell with the next PR. https://trello.com/c/5wutIgRv/1562-move-more-work-in-message-decrypter-and-processor-off-main-threadd

File: SignalServiceKit/src/Messages/TSMessagesManager.m [hunk 23]
```diff
@@ -347,134 +362,172 @@ - (void)processEnvelope:(OWSSignalServiceProtosEnvelope *)envelope
...
-    OWSAssert([NSThread isMainThread]);
+    OWSAssert(envelope);
+    OWSAssert(transaction);
+    OWSAssert([TSAccountManager isRegistered]);
 
-    @synchronized(self) {
-        TSStorageManager *storageManager = [TSStorageManager sharedManager];
-        NSString *recipientId = preKeyEnvelope.source;
-        int deviceId = preKeyEnvelope.sourceDevice;
+    DDLogInfo(@"%@ received envelope: %@", self.tag, [self descriptionForEnvelope:envelope]);
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalServiceKit/src/Messages/TSMessagesManager.m]: Done.

File: SignalServiceKit/src/Messages/TSMessagesManager.m [hunk 24]
```diff
@@ -347,134 +362,172 @@ - (void)processEnvelope:(OWSSignalServiceProtosEnvelope *)envelope
...
+                  transaction:(YapDatabaseReadWriteTransaction *)transaction
+{
+    OWSAssert(envelope);
+    OWSAssert(transaction);
+
+    TSInteraction *interaction = [TSInteraction interactionForTimestamp:envelope.timestamp withTransaction:transaction];
+    if ([interaction isKindOfClass:[TSOutgoingMessage class]]) {
+        TSOutgoingMessage *outgoingMessage = (TSOutgoingMessage *)interaction;
+        [outgoingMessage updateWithWasDeliveredWithTransaction:transaction];
     }
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalServiceKit/src/Messages/TSMessagesManager.m]: Done.

File: SignalServiceKit/src/Messages/TSMessagesManager.m [hunk 25]
```diff
@@ -654,16 +713,22 @@ - (void)handleReceivedGroupAvatarUpdateWithEnvelope:(OWSSignalServiceProtosEnvel
...
     OWSAssert(thread);
     OWSAttachmentsProcessor *attachmentsProcessor =
         [[OWSAttachmentsProcessor alloc] initWithAttachmentProtos:dataMessage.attachments
                                                         timestamp:envelope.timestamp
                                                             relay:envelope.relay
                                                            thread:thread
-                                                   networkManager:self.networkManager];
+                                                   networkManager:self.networkManager
+                                                   storageManager:self.storageManager
+                                                      transaction:transaction];
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalServiceKit/src/Messages/TSMessagesManager.m]: One of the many reasons I want to refactor `OWSAttachmentsProcessor`.

File: SignalServiceKit/src/Messages/TSMessagesManager.m [hunk 26]
```diff
@@ -729,13 +799,16 @@ - (void)handleIncomingEnvelope:(OWSSignalServiceProtosEnvelope *)messageEnvelope
         if ([self isDataMessageGroupAvatarUpdate:syncMessage.sent.message]) {
             [recordJob runWithAttachmentHandler:^(TSAttachmentStream *attachmentStream) {
                 TSGroupThread *groupThread =
-                    [TSGroupThread getOrCreateThreadWithGroupIdData:syncMessage.sent.message.group.id];
-                [groupThread updateAvatarWithAttachmentStream:attachmentStream];
-            }];
+                    [TSGroupThread getOrCreateThreadWithGroupIdData:syncMessage.sent.message.group.id
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalServiceKit/src/Messages/TSMessagesManager.m]: You caught a real bug!  Thanks.

File: SignalServiceKit/src/Messages/TSMessagesManager.m [hunk 27]
```diff
@@ -771,49 +844,57 @@ - (void)handleIncomingEnvelope:(OWSSignalServiceProtosEnvelope *)messageEnvelope
             DDLogWarn(@"%@ ignoring unsupported sync request message", self.tag);
         }
     } else if (syncMessage.hasBlocked) {
-        NSArray<NSString *> *blockedPhoneNumbers = [syncMessage.blocked.numbers copy];
-        [_blockingManager setBlockedPhoneNumbers:blockedPhoneNumbers sendSyncMessage:NO];
+        dispatch_async(dispatch_get_main_queue(), ^{
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalServiceKit/src/Messages/TSMessagesManager.m]: Added card for this release: https://trello.com/c/oqrHSOMj/1563-process-incoming-verification-block-and-sent-message-transcripts-synchronously

File: SignalServiceKit/src/Messages/TSMessagesManager.m [hunk 28]
```diff
@@ -771,49 +844,57 @@ - (void)handleIncomingEnvelope:(OWSSignalServiceProtosEnvelope *)messageEnvelope
...
-                                          messageType:TSInfoMessageTypeSessionDidEnd] saveWithTransaction:transaction];
-        }
-    }];
+    OWSAssert(envelope);
+    OWSAssert(dataMessage);
+    OWSAssert(transaction);
+
+    TSContactThread *thread = [TSContactThread getOrCreateThreadWithContactId:envelope.source transaction:transaction];
+
+    if (thread) { // TODO thread should always be nonnull.
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalServiceKit/src/Messages/TSMessagesManager.m]: Done.

**Comment by charlesmchen** [comment]
@michaelkirk Oh I see that you approved this PR.  I'm going to merge it and if you have any remaining questions we can address this in the next PR which touches all of these classes again.

File: SignalServiceKit/src/Messages/OWSMessageReceiver.m (hunk 15 above)
[the remark this answers is not in the document]
Comment (michaelkirk) [reply SignalServiceKit/src/Messages/OWSMessageReceiver.m]: My concern actually wasn't perf - just about code clarify. From a strictly logical standpoint it seems unnecessary. Instead of sticking them in a data structure and enqueuing them later, why don't we just immediately enqueue them?

Is it related to this comment?:

+// We need to use a consistent batch size throughout
 +// the incoming message pipeline (i.e. in the
 +// "decrypt" and "process" steps), or the pipeline
 +// doesn't flow smoothly.

e.g. I can imagine if we had 10 jobs, and handed them off individually as we finished them, the downstream work queue would start with a batch size of 1, and then continue with a batch of 9, which may be be less efficient than waiting a little bit to do a single batch of 10, as you are here.

...but that is just my speculation and correlation across the whole PR - the motivation is not obvious from reading this code locally.

File: SignalServiceKit/src/Messages/TSMessagesManager.m (hunk 19 above)
[the remark this answers is not in the document]
Comment (michaelkirk) [reply SignalServiceKit/src/Messages/TSMessagesManager.m]: Ah, you're right - my mistake.
